eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/Offers.scala

```diff
@@ -197,89 +232,129 @@ object Offers {
 
     val payerSignature: ByteVector64 = records.get[Signature].get.signature
 
-    def checkSignature: Boolean = {
-      val withoutSig = TlvStream(records.records.filter { case _: Signature => false case _ => true }, records.unknown)
-      verifySchnorr("lightning" + "invoice_request" + "payer_signature", rootHash(withoutSig, invoiceRequestTlvCodec).get, payerSignature, payerKey)
+    def isValidFor(offer: Offer): Boolean = {
+      val amountOk = offer.amount match {
+        case Some(offerAmount) =>
+          val baseInvoiceAmount = offerAmount * quantity
+          amount.forall(baseInvoiceAmount <= _)
+        case None => amount.nonEmpty
+      }
+      amountOk &&
+        offer.offerId == offerId &&
+        offer.chains.contains(chain) &&
+        offer.quantityMin.forall(min => quantity_opt.nonEmpty && min <= quantity) &&
+        offer.quantityMax.forall(max => quantity_opt.nonEmpty && quantity <= max) &&
+        quantity_opt.forall(_ => offer.quantityMin.nonEmpty || offer.quantityMax.nonEmpty) &&
+        offer.features.areSupported(features) &&
+        checkSignature()
     }
 
-    def encode(): String = Bech32WithoutChecksum.encode("lnr", invoiceRequestCodec, this)
+    def checkSignature(): Boolean = {
+      verifySchnorr(InvoiceRequest.signatureTag, rootHash(removeSignature(records), invoiceRequestTlvCodec), payerSignature, payerKey)
+    }
+
+    def encode(): String = Bech32WithoutChecksum.encode(InvoiceRequest.hrp, invoiceRequestCodec, this)
+
+    override def toString: String = encode()
   }
 
   object InvoiceRequest {
-    def apply(offer: Offer, amount: MilliSatoshi, quantity: Long, features: Features[InvoiceFeature], payerKey: PrivateKey, chain: ByteVector32 = Block.LivenetGenesisBlock.hash): InvoiceRequest = {
+    val hrp = "lnr"
+    val signatureTag: String = "lightning" + "invoice_request" + "payer_signature"
+
+    /**
+     * Create a request to fetch an invoice for a given offer.
+     *
+     * @param offer     Bolt 12 offer.
+     * @param amount    amount that we want to pay.
+     * @param quantity  quantity of items we're buying.
+     * @param features  invoice features.
+     * @param payerKey  private key identifying the payer: this lets us prove we're the ones who paid the invoice.
+     * @param chain_opt (optional) should only be provided when the request is not for bitcoin mainnet.
```

Comment:
Why only not for mainnet? It just adds complexity, I think we should always set it.

Reply:
I agree it would be simpler, but our goal is to have offers as small as possible, so if we can omit that for the most widely use-case, it's a nice win.

Reply:
But this is not an offer, it's an invoice request. We don't need to fit inside a QR code.

Reply:
You're right, I didn't expand the code! I agree, for `invoice_request` we can always include it, that's simpler :+1: